apiserver/facades/agent/uniter/networkinfo.go

```diff
@@ -0,0 +1,458 @@
+// for the retrieval of network information.
+func (n *NetworkInfo) init(tag names.UnitTag) error {
+	var err error
+
+	if n.unit, err = n.st.Unit(tag.Id()); err != nil {
+		return errors.Trace(err)
+	}
+
+	if n.app, err = n.unit.Application(); err != nil {
+		return errors.Trace(err)
+	}
+
+	if n.bindings, err = n.app.EndpointBindings(); err != nil {
+		return errors.Trace(err)
+	}
+
+	if n.defaultEgress, err = n.getModelEgressSubnets(); err != nil {
+		return errors.Trace(err)
+	}
+
+	if n.spaces, err = n.st.AllSpaces(); err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+// getModelEgressSubnets returns model configuration for egress subnets.
+func (n *NetworkInfo) getModelEgressSubnets() ([]string, error) {
+	model, err := n.st.Model()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	cfg, err := model.ModelConfig()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return cfg.EgressSubnets(), nil
+}
+
+// ProcessAPIRequest handles a request to the uniter API NetworkInfo method.
+// TODO (manadart 2019-10-09): This method verges on impossible to reason about
+// and should be rewritten.
+func (n *NetworkInfo) ProcessAPIRequest(args params.NetworkInfoParams) (params.NetworkInfoResults, error) {
+	spaces := set.NewStrings()
+	bindings := make(map[string]string)
+	endpointEgressSubnets := make(map[string][]string)
+
+	result := params.NetworkInfoResults{
+		Results: make(map[string]params.NetworkInfoResult),
+	}
+
+	// For each of the endpoints in the request, get the bound space and
+	// initialise the endpoint egress map with the model's configured
+	// egress subnets. Keep track of the spaces that we observe.
+	for _, endpoint := range args.Endpoints {
+		binding, ok := n.bindings[endpoint]
+		if ok {
+			spaces.Add(binding)
+			bindings[endpoint] = binding
+		} else {
+			if endpoint == "" {
+				bindings[endpoint] = corenetwork.DefaultSpaceName
+			} else {
+				err := errors.NewNotValid(nil, fmt.Sprintf("binding name %q not defined by the unit's charm", endpoint))
+				result.Results[endpoint] = params.NetworkInfoResult{Error: common.ServerError(err)}
+			}
+		}
+		endpointEgressSubnets[endpoint] = n.defaultEgress
+	}
+
+	endpointIngressAddresses := make(map[string]corenetwork.SpaceAddresses)
+
+	// If we are working in a relation context, get the network information for
+	// the relation and set it for the relation's binding.
+	if args.RelationId != nil {
+		endpoint, space, ingress, egress, err := n.getRelationNetworkInfo(*args.RelationId)
+		if err != nil {
+			return params.NetworkInfoResults{}, err
+		}
+
+		spaces.Add(space)
+		if len(egress) > 0 {
+			endpointEgressSubnets[endpoint] = egress
+		}
+		endpointIngressAddresses[endpoint] = ingress
+	}
+
+	var (
+		networkInfos            map[string]state.MachineNetworkInfoResult
+		defaultIngressAddresses []string
+	)
+
+	if n.unit.ShouldBeAssigned() {
+		machineID, err := n.unit.AssignedMachineId()
```

Comment:
it does feel like this block could be refactored into another method with a nice name.

n.machineNetworkInfos()

Reply:
Done.